### DISCUSSION TITLE
Tainting worker nodes?

### DISCUSSION THREAD

**Discussion by maulerjan** [siderolabs/talos#9895] [Q&A] [answered]
Hi all,

I'd like to be able to use:
```
machine:
  nodeTaints:
    - MyTaint: true:NoSchedule
```
on worker nodes.

This is currently not possible because of the NodeRestriction admission controller, which is not configurable.

Disabling this restriction is against CIS hardening guidelines and possible other security implications this might introduce.

Is there a way around this issue other than tainting nodes manually?

We have a declarative setup except calling `talosctl -n ... bootstrap`, which is the only non-declarative step we have in our cluster setup workflow.

Thanks!

**Comment by maulerjan** [comment]
Also, it doesn't seem like the admission controller can be turned off at all.
```
cluster:
  apiServer:
    extraArgs:
      enable-admission-plugins: CertificateApproval,CertificateSigning,CertificateSubjectRestriction,DefaultIngressClass,DefaultStorageClass,DefaultTolerationSeconds,LimitRanger,MutatingAdmissionWebhook,NamespaceLifecycle,PersistentVolumeClaimResize,PodSecurity,Priority,ResourceQuota,RuntimeClass,ServiceAccount,StorageObjectInUseProtection,TaintNodesByCondition,ValidatingAdmissionPolicy,ValidatingAdmissionWebhook
      disable-admission-plugins: NodeRestriction
```

Results in apiserver pods crashing:
```
"command failed" err="[NodeRestriction] in enable-admission-plugins and disable-admission-plugins overlapped"
```

As you can see, there is NO overlap in the config options provided, so Talos probably just throws the config option away and use its own (correct me if I'm wrong).

**Reply by maulerjan** [reply]
Also, the docs are wrong:

> Note: In the default Kubernetes configuration, worker nodes are not allowed to modify the taints

But that's not true, NodeRestriction admission controller is _not_ enabled by default:
https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#which-plugins-are-enabled-by-default

**Reply by smira** [reply]
It is enabled by default in Talos.

**Reply by maulerjan** [reply]
Oh I see, thanks for pointing that out

**Comment by smira** [comment, marked as answer]
You can use kubelet's `registerWithTaints` configuration (and pass it via `machine.kubelet.extraConfig` to taint at the moment the Node is created. Updating taints on the fly is still denied by default.

**Reply by maulerjan** [reply]
Works like a charm! Setting this only on initial node registration is enough for our use-case. Thanks!

**Reply by maulerjan** [reply]
Follow-up question:
You say updating taints on the fly is denied _by default_. But I can't find a way to disable the NodeRestriction controller. Is there any way?

**Reply by smira** [reply]
Looks like we don't support that kind of configuration in Talos, and I'd say that probably it should never be disabled.

**Reply by roman-geraskin-m10** [reply]
why? 🙂

**Comment by funky-byte** [comment]
Hi, little bit misunderstand how to correctly add node with taint as best practice IaC -- not manualy like `kubectl taint ...`
Currently running cluster, now i'm add two purely new nodes with taint `node.kubernetes.io/assignment: infra:NoSchedule`

1. Part of config:
```
machine:
  kubelet:
    extraArgs:
      register-with-taints: node.kubernetes.io/assignment=infra:NoSchedule
  nodeTaints:
    node.kubernetes.io/assignment: infra:NoSchedule
```

2. After install and booting this nodes with errors in dmesg:
```
user: warning: [2025-04-15T09:30:39.759379417Z]: [talos] controller failed {"component": "controller-runtime", "controller": "k8s.NodeApplyController", "error": "2 error(s) occurred:\n\tnodes \"infra2-prod1\" is forbidden: nod
e \"infra2-prod1\" is not allowed to modify taints\n\tclient rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline"}
user: warning: [2025-04-15T09:31:35.628430417Z]: [talos] controller failed {"component": "controller-runtime", "controller": "k8s.NodeApplyController", "error": "2 error(s) occurred:\n\tnodes \"infra2-prod1\" is forbidden: nod
e \"infra2-prod1\" is not allowed to modify taints\n\tclient rate limiter Wait returned an error: rate: Wait(n=1) would exceed context deadline"}
```

3. kubectl describe node:
```
Name:               infra2-prod1
Roles:              <none>
Labels:             beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/os=linux
                    extensions.talos.dev/qemu-guest-agent=9.2.0
                    kubernetes.io/arch=amd64
                    kubernetes.io/hostname=infra2-prod1
                    kubernetes.io/os=linux
                    node.kubernetes.io/assignment=infra
Annotations:        csi.volume.kubernetes.io/nodeid: {"hdd.rbd.csi.ceph.com":"infra2-prod1","ssd.rbd.csi.ceph.com":"infra2-prod1"}
                    extensions.talos.dev/schematic: ce4c980550dd2ab1b17bbf2b08801c7eb59418eafe8f279833297925d67c7515
                    node.alpha.kubernetes.io/ttl: 0
                    node.kubernetes.io/assignment: infra
                    talos.dev/owned-annotations: ["extensions.talos.dev/schematic","node.kubernetes.io/assignment"]
                    talos.dev/owned-labels: ["extensions.talos.dev/qemu-guest-agent","node.kubernetes.io/assignment"]
                    talos.dev/owned-taints: ["node.kubernetes.io/assignment"]
                    volumes.kubernetes.io/controller-managed-attach-detach: true
CreationTimestamp:  Thu, 20 Mar 2025 15:22:10 +0300
Taints:             <none>
Unschedulable:      false
```

Versions:
```
Client:
        Tag:         v1.9.1
        SHA:         348472f9
        Built:
        Go version:  go1.23.4
        OS/Arch:     linux/amd64
Server:
        Tag:         v1.9.4
        SHA:         c863a561
        Built:
        Go version:  go1.23.6
        OS/Arch:     linux/amd64
```

**Reply by smira** [reply]
See the answer https://github.com/siderolabs/talos/discussions/9895#discussioncomment-11506439.

See https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction for specifics. 

tl;dr worker machines can't modify taints in any way.

**Reply by funky-byte** [reply]
```
machine:
  kubelet:
    extraConfig:
      registerNode: true
      registerWithTaints:
        - effect: NoSchedule
          key: node.kubernetes.io/assignment
          value: infra
```
This configuration is work pretty well, thanks for pointing!

**Reply by nrx-ops** [reply]
Not Working for me. :/

**Reply by maulerjan** [reply]
@nrx-ops It only works for **initial** node setup. Once the node is already registered in the cluster, this configuration doesn't have any effect.

**Reply by nrx-ops** [reply]
Thank you! I should have miss something.
I use terraform to generate the machine config.
I tried to apply config with "Reset Talos installation" menu at VM boot. Without this, machineConfig was not updated (Correct behavior).
But the taint never applied after reset. Destroying resources and reapply seems to work.
